### ISSUE TITLE
Performance experiments

### ISSUE THREAD

**Issue by schlegelp** [clbarnes/ncollpyde#39]
Hi Chris.

I'm working on some mesh fixing functions that rely heavily on raycasting and had Claude try a few experiments to push performance in `ncollpyde`.

The whole thing was sparked because I noticed that Embree (via `open3d`'s `RaycastingScene`, the same kernel `pyembree` wraps) on identical geometry, Embree is a lot faster — especially for distance queries:

| query | ncollpyde | Embree | gap |
|---|---|---|---|
| `contains` (3 rays, threaded) / `compute_occupancy` | 0.049 µs/pt | 0.014 µs/pt | 3.5× |
| `distance(signed=False)` (threaded) / `compute_distance` | 12.20 µs/pt | 0.868 µs/pt | **14×** |

I went looking for where that goes, and found five things. Four are ordinary code or build-config issues. The fifth accounts for most of the distance gap on its own, and I think it's the interesting one. Wanted to hear your thoughts on whether/how to incorporate any of that in `ncollpyde`.

Below is Claude's long (sorry) summary:

---

## 1. `intersections(threads=True)` is slower than `threads=False`

`intersections_many_threaded` takes `Vec<Vec<Precision>>` and returns `(Vec<u64>, Vec<Vec<Precision>>, Vec<bool>)`, so `main.py` calls `.tolist()` on both inputs and rebuilds arrays from lists afterwards — one Python float object per coordinate, in both directions. The marshalling costs more than the parallelism saves. On my 14-core MacBook Pro:

| | before | after |
|---|---|---|
| `threads=False` | 0.616 µs/seg | 0.578 µs/seg |
| `threads=True` | **0.712 µs/seg** | **0.054 µs/seg** |

The docstring already concedes this ("you may not see the performance increase as with other methods"), but on 14 cores the "parallel" path was losing to serial outright.

The fix is mechanical: give it the same `PyReadonlyArray2` signature as `intersections_many` and parallelise with `Zip::par_map_collect`, exactly as `contains` and `distance` already do. Collect one `Option<(Point, bool)>` slot per segment so both paths stay input-ordered, then compact the misses out serially — that compaction is trivial next to the ray casts.

**13× faster than the old threaded path, and 10.7× faster than serial** instead of slower.

Incidental bug this fixes: the threaded path returned `uint32` indices while the serial path returned `uint64`, contradicting the declared
`Tuple[NDArray[np.uint64], ...]`. Unifying the paths makes both `uint64`.

## 2. `distance()` has no `max_distance` cut-off

`distance()` always finds the true closest point. Most callers only need the near field ("is this within X nm of the mesh?"), and parry can prune the traversal hard if you give it an upper bound.

Adding an optional `max_distance` is purely additive — points beyond the cut-off report `inf`, keeping their sign so interior points read as `-inf` and callers can still tell inside from outside:

| `max_distance` | µs/pt | vs unbounded |
|---|---|---|
| none | 126.4 | 1.0× |
| 1.0 | 108.0 | 1.2× |
| 0.5 | 40.7 | 3.1× |
| 0.1 | 2.97 | 42.6× |
| 0.01 | 0.628 | **201×** |

(unit-radius sphere, so a cut-off of 1.0 barely prunes — most query points are already nearer than that.)

One implementation trap worth recording: `PointQuery::project_local_point_with_max_dist` has a **default** trait implementation that projects fully and *then* discards the result if it's too far — which would buy nothing. `TriMesh` overrides it and feeds the bound into the BVH traversal properly (the unbounded path is implemented as `..._with_max_dist(point, solid, Real::MAX)`), so the saving is real, but it's worth
checking rather than assuming for any other shape.

## 3. `points()` / `faces()` allocate a `Vec` per element

Both build a `Vec<Vec<T>>` — one heap allocation per vertex/face — before `PyArray2::from_vec2`. `TriMesh` stores these contiguously, so a single flat buffer plus one `Array2::from_shape_vec` is a straight memcpy:

| | before | after |
|---|---|---|
| `points` | 0.024 µs/vertex | 0.002 µs/vertex (**15×**) |
| `faces` | 0.022 µs/face | 0.001 µs/face (**44×**) |

Only affects property access, but it's noticeable on large meshes and the change is a few lines.

## 4. No `[profile.release]`

`Cargo.toml` has no release profile, so LTO is off and `codegen-units = 16`. The query loops reach parry's BVH traversal through generic trait methods across a crate boundary, which is exactly the case fat LTO helps.

```toml
[profile.release]
lto = "fat"
codegen-units = 1
```

Real but modest: **~15% on `contains`, ~11% on `intersections`, ~3% on `distance`**. Free, and it only costs build time.

## 5. `enhanced-determinism` costs ~10× and appears to buy nothing here

This is the big one, and it's a one-line change.

`parry3d-f64` is built with `enhanced-determinism`. That expands to `["simba/libm_force", "indexmap"]`, and `libm_force` routes every `sqrt` through a software implementation instead of the hardware instruction. Distance queries are almost entirely `sqrt`:

| query (serial) | determinism on | determinism off |
|---|---|---|
| `distance` | 118.4 µs/pt | **12.06 µs/pt** |
| `distance`, `max_distance=0.01` | 0.628 µs/pt | 0.059 µs/pt |
| `contains` (3 rays) | 0.414 µs/pt | 0.379 µs/pt |
| `intersections` | 0.578 µs/pt | 0.485 µs/pt |

**9.8× on distance queries**, ~20% elsewhere. For reference, that single flag is most of the Embree gap in the opening table: threaded `distance` goes from 12.20 to 1.10 µs/pt, against Embree's 0.868 — a 14× gap becomes 1.27×, in f64 against Embree's f32.

### It doesn't appear to change any results

The obvious objection is that this is exactly the trade the flag exists to make. So I tested it. I captured 32 output arrays from both builds — two meshes (an icosphere and this repo's `teapot.stl`), both threading modes, signed / unsigned / cut-off distance, containment, intersection indices, locations and backface flags, plus the generated rays and vertex buffers — and compared them as raw bytes.

**All 32 were bit-for-bit identical.** Control: the same timing snippet measures 60.52 µs/pt with the flag on and 6.06 µs/pt with it off, so the two builds were unambiguously different binaries.

There's a mechanical reason:

- **The `indexmap` half never runs.** Only the voxels, mesh-boolean,  contact-manifold and convex-hull code uses parry's hash-map shim, and none of  ncollpyde's query paths touch it. Even where it *is* used, the non-determinism  fallback is `foldhash`'s `FixedState` — a fixed seed, not a randomised one — so there's no run-to-run variation either way.
- **The `libm_force` half only affects `sqrt` here.** The query paths (ray/triangle, point/triangle, AABB) are dot products, cross products, comparisons and `sqrt`. No transcendentals — which are the operations that actually differ between implementations. IEEE-754 *requires* `sqrt` to be correctly rounded, so software and hardware must agree exactly. The flag is paying for a software `sqrt` loop to guarantee something the hardware already guarantees.

### What this does *not* prove

All of the above is one machine, arm64. Cross-platform reproducibility is precisely what the flag is for, and precisely what a single machine can't demonstrate. The IEEE argument for `sqrt` is platform-independent and should hold on x86-64, but "should" is doing real work in that sentence.

That gap is cheap to close — CI already builds on macOS, Windows and Ubuntu, so running the same capture-and-compare across all three would settle it with evidence. Worth doing before dropping the flag, and worth keeping afterwards as a regression test: the conclusion holds only while the query paths stay free of transcendentals and away from parry's hash-map shim, so it would need revisiting if the API grew
into mesh booleans or contact manifolds.

### Note: `simd-stable` is not a separate option

`Cargo.toml` doesn't enable `simd-stable`, which looks like an obvious win — parry's QBVH/BVH traversal is built around 4-wide SIMD AABB tests, and without the feature `SimdReal` is a scalar-emulated type. But it can't be combined with the flag above; parry has a hard guard, on 0.13.3 and 0.25.3 alike:

```rust
#[cfg(all(feature = "simd-is-enabled", feature = "enhanced-determinism"))]
std::compile_error!(
    "SIMD cannot be enabled when the `enhanced-determinism` feature is also enabled."
);
```

And measured on top of a determinism-free build, it's worth very little: `distance` 12.06 → 11.28 µs/pt (1.07×), `intersections` 0.485 → 0.472, and `contains` actually gets *slower*, 0.379 → 0.413. So the choice isn't "SIMD vs reproducibility" — SIMD can be set aside, and the question is just whether the 10× is worth the guarantee. (This is arm64, where `f64x4` is two 128-bit NEON registers; an x86 box with AVX2 might rank SIMD better, but it would have to beat 1.07× by a lot to matter.)

---

## Also worth mentioning

**No GIL release anywhere.** The GIL is held for the whole of every query, so nothing else in the process runs during a multi-second `distance()` call — relevant to anyone calling ncollpyde from a thread pool. Wrapping the query bodies in `py.detach()` (`allow_threads` before pyo3 0.26) fixes it: a Python thread measured at 92.6% of its uncontended rate during a query, and it composes fine with the existing
`PyReadonlyArray2` borrows.

**Three things blocked building at all** on a current toolchain, all pre-existing:

- `[dependencies.ndarray]` is declared with no version. Cargo 1.91 makes that a hard error — the comment beside it predicted exactly this ("may be an error in future").
- `pyproject.toml` has `version` commented out, which current maturin rejects; it needs `dynamic = ["version"]`.
- `cargo-release` is in `dev-dependencies`. It's a CLI, not a library — cargo ignores it as a dependency ("missing a lib target") but still resolves its tree, pulling in a `time` version that no longer compiles. Removing it is what let `cargo test` run.

The Rust test module is also broken independently on newer parry (`TriMesh::new` returns a `Result`).

---

## Method and caveats

- 14-core Apple silicon (arm64), rustc 1.91, Python 3.11.8, **parry 0.25.3**, pyo3 0.27.
- Geometry: `trimesh.creation.icosphere(subdivisions=6)` — 40,962 vertices, 81,920 faces. Query points uniform in [-1.5, 1.5]³, so a mix of interior, near-surface and far-exterior. Distance over 20k points, containment over 200k, intersections over 100k segments. Best of three runs after a warm-up.
- **These are measured against parry 0.25.3, not the 0.13 currently pinned on `main`.** The findings are structural and should carry over, but the absolute numbers won't.
- Run-to-run variance is about 7% — the same unbounded `distance` query measured 118.4 and 126.4 µs/pt within one session. Treat anything under ~1.1× as noise. 
- Embree comparison caveats: `open3d`'s `RaycastingScene` is f32 against ncollpyde's f64, parallelises internally (so it's compared against ncollpyde's threaded numbers), and `compute_occupancy` uses a different algorithm from the multi-ray consensus heuristic — it's a reference point, not a like-for-like kernel benchmark.
- One machine, one architecture throughout.

**Comment by clbarnes** [comment]
Thanks for looking into these! I appreciate the value you've got from the LLM during your investigation and would be happy to accept individual PRs on any of these points. Certainly there's a lot left in this lib from a much less experienced Chris-as-rustacean ;) . As a maintainer, I do have a preference for interacting with you and your understanding of the findings rather than raw LLM output.

**Comment by schlegelp** [comment]
Fully understand. For what it's worth: the preamble is obviously me. As for the report/summary: I don't think I would have done a better job by rewording it. I should have made more of an effort to ask specific questions rather than dump this on you though!

From my point of view these are all easily separable:
1. Fix threading in `intersection`
2. Adding `max_distance` to `distance()` 
3. Flatten `points()` / `faces()`
4. Add `[profile.release]`
5. Drop `enhanced-determinism`

I would have suggested making 5 separate PRs for this - easier to review than a single monolithic PR?

The one thing that would be nice to get in before I do that is #38 so that there a clean base on which I can layer the rest. In your latest message in that PR you had moved it into a separate branch and were close on getting it to work?

**Comment by clbarnes** [comment]
Yes; I imagine all the dependencies have moved on somewhat since then and may have (hopefully) solved some of the problems. There's also #36 which already addresses some of the raised points about Vec<Vec<_>>, array copies, and so on.
